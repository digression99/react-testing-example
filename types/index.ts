export type User = {
  username: string
  email: string
  profileUrl: string
}
