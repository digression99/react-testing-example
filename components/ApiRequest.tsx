import {useEffect, useState} from "react";
import {User} from "../types";
import {Box} from "@chakra-ui/react";
import Image from "next/image";

export const ApiRequest = () => {
  const [user, setUser] = useState<User>()
  const [loading, setLoading] = useState(false)

  useEffect(() => {
    setLoading(true)

    fetch('/api/user')
      .then(res => res.json())
      .then(user => {
        setUser(user)
        setLoading(false)
      })
  }, [])

  return (
    <Box>
      {loading && (<Box>Loading...</Box>)}
      {user && (
        <Box>
          username: {user.username}
          email: {user.email}
          profilePicture: <Image src={user.profileUrl} alt={'profile-pic'} width={100} height={100} />
        </Box>
      )}
    </Box>
  )
}
