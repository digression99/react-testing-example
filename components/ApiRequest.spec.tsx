import {rest } from 'msw'
import { setupServer } from 'msw/node'
import { faker } from '@faker-js/faker'

import { render, screen } from '@testing-library/react'
import { ApiRequest } from "./ApiRequest";
import {User} from "../types";

faker.seed(500)

const server = setupServer(
  rest.get('/api/user', (req, res, ctx) => {
    const fakeUser: User = {
      username: 'test username',
      email: 'test email',
      profileUrl: faker.image.people()
    }

    return res(
      ctx.json(fakeUser)
    )
  })
)

describe('<ApiRequest />', () => {
  beforeAll(() => server.listen())
  afterEach(() => server.resetHandlers())
  afterAll(() => server.close())

  test('renders correctly', () => {
    const { container } = render(<ApiRequest />)
    expect(container).toMatchSnapshot()
  })

  test('show the user component', async () => {
    render(<ApiRequest />)
    const usernameText = await screen.findByText(/test username/i)
    expect(usernameText).toBeInTheDocument()
  })
})