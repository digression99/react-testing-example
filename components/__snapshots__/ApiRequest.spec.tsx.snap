// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ApiRequest /> renders correctly 1`] = `
<div>
  <div
    class="css-0"
  >
    <div
      class="css-0"
    >
      Loading...
    </div>
  </div>
</div>
`;
